File: backend/methods/false_position.py
from .utils import evaluate_function

def false_position_method(func_str, a, b, tol=1e-6, max_iter=100):
    """
    Solves for the root of a function using the False Position (Regula Falsi) Method.
    """
    steps = []
    
    fa = evaluate_function(func_str, a)
    fb = evaluate_function(func_str, b)
    
    if fa * fb >= 0:
        return {
            "success": False,
            "message": "Function must have opposite signs at a and b (f(a) * f(b) < 0).",
            "root": None,
            "steps": steps
        }
    
    x_prev = a
    for i in range(1, max_iter + 1):
        # False position formula
        c = (a * fb - b * fa) / (fb - fa)
        fc = evaluate_function(func_str, c)
        
        error = abs(c - x_prev) if i > 1 else abs(b - a)
        
        step_data = {
            "iteration": i,
            "a": a,
            "b": b,
            "c": c,
            "f_c": fc,
            "error": error
        }
        steps.append(step_data)
        
        if error < tol or fc == 0:
            return {
                "success": True,
                "message": f"Converged to root after {i} iterations.",
                "root": c,
                "steps": steps
            }
            
        if fa * fc < 0:
            b = c
            fb = fc
        else:
            a = c
            fa = fc
            
        x_prev = c
            
    return {
        "success": False,
        "message": f"Did not converge within {max_iter} iterations.",
        "root": x_prev,
        "steps": steps
    }

File: worker/src/methods/falsePosition.ts
import { evaluateFunction } from "../lib/exprEval";

export interface FalsePositionStep {
  iteration: number;
  a: number;
  b: number;
  c: number;
  f_c: number;
  error: number;
}

export function falsePositionMethod(
  funcStr: string,
  a: number,
  b: number,
  tol = 1e-6,
  maxIter = 100
) {
  const steps: FalsePositionStep[] = [];

  let fa = evaluateFunction(funcStr, a);
  let fb = evaluateFunction(funcStr, b);

  if (fa * fb >= 0) {
    return {
      success: false,
      message: "Function must have opposite signs at a and b (f(a) * f(b) < 0).",
      root: null,
      steps,
    };
  }

  let xPrev = a;

  for (let i = 1; i <= maxIter; i++) {
    const c = (a * fb - b * fa) / (fb - fa);
    const fc = evaluateFunction(funcStr, c);

    const error = i > 1 ? Math.abs(c - xPrev) : Math.abs(b - a);

    steps.push({ iteration: i, a, b, c, f_c: fc, error });

    if (error < tol || fc === 0) {
      return {
        success: true,
        message: `Converged to root after ${i} iterations.`,
        root: c,
        steps,
      };
    }

    if (fa * fc < 0) {
      b = c;
      fb = fc;
    } else {
      a = c;
      fa = fc;
    }

    xPrev = c;
  }

  return {
    success: false,
    message: `Did not converge within ${maxIter} iterations.`,
    root: xPrev,
    steps,
  };
}
